File: Erlang/knight_tour.erl
-module(knight_tour).
-export([solve/3]).

get_length(L, Size) ->
  [ M || M <- lists:map( fun([X, Y]) -> [length(neighbours(X, Y, Size)), [X, Y]] end, L) ].


get_move(L) ->
  lists:last( lists:min(lists:filter( fun([H|_T]) -> H > 0 end, L)) ).

solve(X, Y, Size) ->
  solve(X, Y, Size, neighbours(X, Y, Size), [[X, Y]]).


solve(X, Y, Size, _Neighbours, Moves) ->

  NewNeighbours = neighbours(X, Y, Size) -- Moves,

  case length(NewNeighbours) == 0 of
    true -> lists:reverse(Moves);
    false -> [X1, Y1] = get_move(get_length(NewNeighbours, Size)),
             solve(X1, Y1, Size, NewNeighbours, lists:append(Moves, [[X1, Y1]]))
  end.

  

is_valid_point([X, Y], Size) ->
  is_integer(X) andalso X >= 0 andalso X < Size andalso
  is_integer(Y) andalso Y >= 0 andalso Y < Size.


neighbours(X, Y, Size) ->
  Candidate = [[X+1, Y-2], [X+1, Y+2], [X+2, Y+1], [X+2, Y-1], [X-1, Y-2], [X-1, Y+2], [X-2, Y+1], [X-2, Y-1]],
  [R || [_X, _Y] = R <- Candidate, is_valid_point(R, Size)].

File: Python/knight_tour.py
import sys


def is_valid_point(point, size):
    x = isinstance(point[0], int) and point[0] >= 0 and point[0] < size
    y = isinstance(point[1], int) and point[1] >= 0 and point[1] < size
    return x and y


def get_neighbours(x, y, size):
    candidate = [[x+1, y-2], [x+1, y+2],
                 [x+2, y+1], [x+2, y-1],
                 [x-1, y-2], [x-1, y+2],
                 [x-2, y+1], [x-2, y-1]]
    return [c for c in candidate if is_valid_point(c, size)]


def get_next_move(point, size, moves):
    lengths = []
    get_neighbours(point[0], point[1], size)

    for n in get_neighbours(point[0], point[1], size):
        if n not in moves:
            lengths.append([len(get_neighbours(n[0], n[1], size)), n])

    if lengths:
        next_move = sorted(lengths)[0][1]
        return next_move[0], next_move[1]
    else:
        return None, None


def solve(x, y, size, moves=[]):
    moves.append([x, y])
    neighbours = get_neighbours(x, y, size)
    neighbours = [n for n in neighbours if n not in moves]
    next_x, next_y = get_next_move([x, y], size, moves)

    if neighbours and (next_x is not None and next_y is not None):
        return solve(next_x, next_y, size, moves)
    else:
        for m in reversed(moves):
            print m


solve(int(sys.argv[1]), int(sys.argv[2]), int(sys.argv[3]))
